torch/nested/_internal/nested_tensor.py

```diff
@@ -3,23 +3,224 @@
+    # canonical or not has no bearing on behavior. In fact, one should consider
+    # all nested ints (given that coeff=1) and vec that are associated with the
+    # same equiv set to be fungible when dealing with the registry.
+    #
+    # One consequence of having canonical vec is that vec pointed to by
+    # nt.shape[ragged_dim].node.nested_int_vec() is not necessarily the same
+    # Tensor as NT's ragged_source. This shouldn't matter though as per the
+    # above.
+    #
+    # Example:
+    # --------
+    # In the diagram below, arrows represent the direction of ownership. j0_0
+    # and j0_1 are in the same equiv set, but are different instances. vec0 and
+    # j0_0 are the canonical vec and nested int for equiv_set0.
+    #
+    #  symint       vec         equiv_set       cached metadata
+    # --------------------------------------------------------------------------
+    #  2*j0 -----\                        /---- weak vec0, weak j0_0
+    #             v                      | /--> id: int = 0
+    #  j0_0 ------> vec0 -----> equiv_set0 ---> sum_vec: Optional[SymInt]
+    #  j0_1 ------^         /-^            \--> vec_{cuda,cpu}: Optional[Tensor]
+    #                      /
+    #               vec1--/               /---- weak vec2, weak j1_0
+    #                                    | /--> id: int = 1
+    #  j1_0 ------> vec2 -----> equiv_set1 ---> ...
+    #
+    # Details on canonical vec and nested int lifetimes:
+    # --------------------------------------------------
+    # Throughout the lifetime of a equiv set, the canonical vec may change.
+    # If we detect that the canonical vec is no longer alive, the next time
+    # someone asks us for a nested int for a vec in that equiv set we will make
+    # that vec the canonical vec. The canonical vec may be on cpu or cuda.
+    #
+    # If canonical vec and nested int are alive, nested int's vec must be the
+    # canonical vec. This implies that if canonical is not alive, then the
+    # nested int is also not alive. It is possible for canonical to be alive,
+    # but for nested int to be dead. This happens when the nested int is only
+    # alive in cpp, but its corresponding python object has died.
+    #
+    # When you have tracing subclass tensors as vec
+    # ---------------------------------------------
+    # During tracing, we may be dealing with vec which are subclasses used
+    # for tracing, e.g. FakeTensor and FunctionalTensor. Though these tensors
+    # don't actually need equiv set for the purpose of checking equality, we
+    # the concepts of equiv sets are still useful so that we can reuse the logic
+    # in eager to access the cpu and cuda variants from one another. Note that
+    # that since each equiv set has its own cache, FakeTensor must be in a
+    # different equiv set from real tensor.
+    #
+    def __init__(self):
+        self._equiv_set_counter = 0
+        self._equiv_sets = WeakTensorKeyDictionary()
```

Comment:
nit: document the mapping modeled here. My understanding is it's something like `tensor weakref -> dict of properties associated with the equiv set` (including id, canonical vec, canonical nested int, weakrefs to all vecs, etc.)